[bug] Multilanguage: Save as Copy should not transfer the associations to the new item

infograf768 opened the pull request:
See https://github.com/joomla/joomla-cms/issues/8907

Test: 
install a basic multilingual site with the staging branch.
Enable item associations in the language filter plugin.
Associate some items (article, categories, contact, etc.)

Before patch
When using "Save as Copy" when editing an item with associations, the associations are transferred to the newly created copy instead of letting the original item with its associations.

After patch, the original item is untouched.

infograf768 committed:
multilanguage: Save as Copy should not transfer the associations to the
new item

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on 3f21121b94be9ebfa0cdf519880c2b7cad3f79fc<br /><br />I can confirm this problem in articles, categories and content. I can't confirm in the menus.
And can confirm the patch solves the problem in articles, categories and content.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8918">issues.joomla.org/joomla-cms/8918</a>.</sub>

infograf768 committed:
Using similar code as for menu items

infograf768 commented:
@andrepereiradasilva
You are right for menus. It was not needed there as there is an override taking care of that.
Modified the code here to use same as for menu items.
Just changed to `$data['associations'] = array();` instead of `$data['associations'] = '';`

fontanil commented:
I have tested this item :white_check_mark: successfully on 242a06907381a16a35ca00051ad671c71d781fb8<br /><br />Test OK, thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8918">issues.joomla.org/joomla-cms/8918</a>.</sub>

infograf768 commented:
@andrepereiradasilva
This needs a new test to get RTC

andrepereiradasilva commented:
I have tested this item :white_check_mark: successfully on 242a06907381a16a35ca00051ad671c71d781fb8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8918">issues.joomla.org/joomla-cms/8918</a>.</sub>

infograf768 commented:
RTC. Thanks for testing.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8918">issues.joomla.org/joomla-cms/8918</a>.</sub>

infograf768 commented:
This should be in 3.5.0. Thanks

wilsonge commented on libraries/legacy/controller/form.php:
```diff
@@ -661,6 +661,12 @@ public function save($key = null, $urlVar = null)
 			$data[$key] = 0;
+
+			// Reset multilingual associations if necessary
+			if (isset($data['associations']) && array_sum($data['associations']) != 0)
```
Just doing code review - is there a reason we are using `array_sum($data['associations']) != 0` rather than `count($data['associations']) > 0` which would be better for performance

infograf768 replied:
Because `count($data['associations'])` returns the number of the variables (here Content Languages) and not the values, if any.
I.e. when associations are implemented in the language filter, when using `count` the figure will always be equal to the number of Content Languages while `array_sum` will return the sum of the ids associated.

Exemple:
On a site with 6 Content Languages, one article with only one association with another article which id is 150 will return
```
using count

int 6

using array_sum

int 150
```

If the article is associated with  2 articles (each one in a different Content Language), and the second article id is 79, we will have 150+79 = 229
```
using count

int 6

using array_sum

int 229
```

If the article is not associated to any other article we will get
```
using count

int 6

using array_sum

int 0
```
